### DISCUSSION TITLE
PostgreSQL: should the preferred time be timestamptz?

### DISCUSSION THREAD

**Discussion by brecht-derooms** [prisma/prisma#18275] [Q&A] [answered]
I recall a big change in npgsql related to timestamp vs timestamptz where they made the choice to require DateTimes with UTC timezone to be written to timestamptz columns. https://www.npgsql.org/doc/types/datetime.html

Besides that, it seems to be the preferred default according to this: https://dba.stackexchange.com/questions/265550/any-downside-to-using-timestamp-without-time-zone-if-utc-is-ok-at-all-times.

Postgres also advises storing UTC timestamps in https://wiki.postgresql.org/wiki/Don%27t_Do_This#:~:text=might%20be%20suitable.-,Don't%20use%20timestamp%20(without%20time%20zone)%20to%20store,timestamp%20with%20time%20zone%20instead.

Having experienced working with npgsql (and witnessed that change) I wondered what would be generated here. Is there any reason to favor timestamp over timestamptz columns since I assume most people would be storing UTC timestamps?

**Comment by nurul3101** [comment, marked as answer]
Hello @brecht-derooms 👋 

You can favour timestamp over timestamptz if you are dealing with only one timezone.

In my opinion, `timestamptz` provides a more consistent way to store and retrieve timestamps across different time zones, as it automatically converts timestamps to the appropriate time zone based on the session time zone setting. One other alternative that people use is that they store just the timestamp in one column and store the timezone in another text column. Ultimately, the choice between timestamp and timestamptz depends on your specific requirements and use case.

**Reply by brecht-derooms** [reply]
Hi @nurul3101, I am confused, it seems we agree that timestamptz is a more consistent way and will be more beneficial to most of your users so I assumed it was a more sensible default. Indeed, there are use cases for timestamp for users who know the difference. 

I just wanted to be kind and let you know that Prisma seems to take another stance than most ORMs or the general Postgres advice which might be less than ideal for your users. At least in my tests, a simple DateTime without annotations results in the following migration:
```
ALTER TABLE "categories" ALTER COLUMN "created_at" SET DATA TYPE TIMESTAMP(3);
```
So I personally annotate all my DateTime's with `@db.Timestamptz(6)` to get :
```
ALTER TABLE "categories" ALTER COLUMN "created_at" SET DATA TYPE TIMESTAMPTZ(6);
```

**Reply by nurul3101** [reply]
@brecht-derooms You are correct, by default, a simple DateTime would result in migration with `TIMESTAMP(3);` and won't include a timestamp with timezone `TIMESTAMPTZ(6)`. I guess this is the decision our engineering team made by not assuming a timezone by default. Can you share other ORMs which went the other way? I can discuss this internally.

**Reply by brecht-derooms** [reply]
**Sequelize** is one: https://sequelize.org/docs/v7/other-topics/other-data-types/
**ActiveRecord**: I recall this post on rails' ActiveRecord, https://medium.com/@frodsan/activerecord-better-native-types-mappings-for-postgresql-b5391d14ea68. Apparently, they changed it in 2021 https://github.com/rails/rails/pull/41084

Others use types or other means to guide their users in the 'right' direction. For example in **Npgsql (.NET)** or **SeaORM** they prohibit you from inserting a rust DateTime<Utc> or a .NET DatetTime with an UTC timezone into a field backed by a Postgres "DateTime without timezone" column. That's their way of making it clear that if your intention is to store UTC in the database, you are probably better off with 'with timezone'. In **Ecto (elixir)** apparently from what I can see deprecated the regular 'datetime' type with warning: "the :datetime type in migrations is deprecated, please use :utc_datetime or :naive_datetime instead.". The new alternatives are highly descriptive which is their way of protecting users to chose the less optimal default. 

Of course, given that Prisma is defined in GraphQL terms and has clients in many languages, that's an entirely different story.

**Reply by nurul3101** [reply]
Thanks for sharing the links 🙌 
I'll share this internally so that we can investigate how we would like to treat this use case 👍
